pkg/yurthub/util/fs/store.go

```diff
@@ -0,0 +1,288 @@
+			if err != nil {
+				return err
+			}
+
+			switch {
+			case info.Mode().IsDir():
+				if path == rootDir {
+					return nil
+				}
+				dirs = append(dirs, path)
+			case info.Mode().IsRegular():
+				files = append(files, path)
+			default:
+				// something like symbolic link, currently ignore it
+			}
+			return nil
+		})
+	} else {
+		infos, err := os.ReadDir(rootDir)
+		if err != nil {
+			return nil, err
+		}
+		for i := range infos {
+			switch {
+			case infos[i].Type().IsDir():
+				dirs = append(dirs, filepath.Join(rootDir, infos[i].Name()))
+			case infos[i].Type().IsRegular():
+				files = append(files, filepath.Join(rootDir, infos[i].Name()))
+			default:
+				// something like symbolic link, currently ignore it
+			}
+		}
+	}
+
+	switch mode {
+	case ListModeDirs:
+		sort.Strings(dirs)
+		return dirs, nil
+	case ListModeFiles:
+		sort.Strings(files)
+		return files, nil
+	default:
+		return nil, fmt.Errorf("unknown mode %s", mode)
+	}
+}
+
+// DeleteFile will delete file at path.
+// If the path does not exist, do nothing.
+// If the path exists but is not a regular file, return ErrIsNotFile.
+func (fs *FileSystemOperator) DeleteFile(path string) error {
+	if !IfExists(path) {
+		return nil
+	}
+
+	if ok, err := IsRegularFile(path); !ok || err != nil {
+		if !ok {
+			return ErrIsNotFile
+		}
+		return err
+	}
+
+	return os.RemoveAll(path)
+}
+
+// DeleteDir will delete directory at path. All files and subdirs will be deleted.
+// If the path does not exist, do nothing.
+// If the path does not a directory, return ErrIsNotDir.
+func (fs *FileSystemOperator) DeleteDir(path string) error {
+	if !IfExists(path) {
+		return nil
+	}
+
+	if ok, err := IsDir(path); !ok || err != nil {
+		if !ok {
+			return ErrIsNotDir
+		}
+		return err
+	}
+
+	return os.RemoveAll(path)
+}
+
+// CreateDir will create the dir at path. If the parent dir of this path does not
+// exist, it will be created first.
+// If the path has already existed and is not dir, return ErrIsNotDir.
+// If the path has already existed and is a dir, return ErrExists.
+func (fs *FileSystemOperator) CreateDir(path string) error {
+	if IfExists(path) {
+		if ok, err := IsDir(path); !ok && err == nil {
+			return ErrIsNotDir
+		}
+		return ErrExists
+	}
+
+	if err := os.MkdirAll(path, 0755); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CreateFile will create the file at path with content. If its parent dir does not
+// exist, it will be created first.
+// If the path has already existed and is not regular file, return ErrIsNotFile.
+// If the path has already existed and is regular file, return ErrExists.
+func (fs *FileSystemOperator) CreateFile(path string, content []byte) error {
+	// check the path
+	if IfExists(path) {
+		if ok, err := IsRegularFile(path); !ok && err == nil {
+			return ErrIsNotFile
+		}
+		return ErrExists
+	}
+
+	// ensure the base dir
+	dir := filepath.Dir(path)
+	if _, err := os.Stat(dir); err != nil {
```

Comment:
line 214: `IfExists` has already called `os.Stat` to check existence of path, so we should not call `os.Stat()` twice here.

Reply:
The first IfExists checks if the file exists, and the second check if the dir of this file exists. We need to create the dir if it does not exist before create the file.